source/extensions/tracers/opentelemetry/tracer.cc

```diff
@@ -0,0 +1,138 @@
+  flush_timer_->enableTimer(std::chrono::milliseconds(flush_interval));
+}
+
+void Tracer::flushSpans() {
+  ExportTraceServiceRequest request;
+  // A request consists of ResourceSpans.
+  ::opentelemetry::proto::trace::v1::ResourceSpans* resource_span = request.add_resource_spans();
+  ::opentelemetry::proto::trace::v1::InstrumentationLibrarySpans* instrumentation_library_span =
+      resource_span->add_instrumentation_library_spans();
+  for (const auto& pending_span : span_buffer_) {
+    (*instrumentation_library_span->add_spans()) = pending_span;
+  }
+  tracing_stats_.spans_sent_.add(span_buffer_.size());
+  if (!exporter_->log(request)) {
+    // TODO: should there be any sort of retry or reporting here?
+    ENVOY_LOG(trace, "Unsuccessful log request to OpenTelemetry trace collector.");
+  }
+  span_buffer_.clear();
+}
+
+void Tracer::sendSpan(::opentelemetry::proto::trace::v1::Span& span) {
+  span_buffer_.push_back(span);
+  const uint64_t min_flush_spans =
+      runtime_.snapshot().getInteger("tracing.opentelemetry.min_flush_spans", 5U);
+  if (span_buffer_.size() >= min_flush_spans) {
+    flushSpans();
+  }
+}
+
+Tracing::SpanPtr Tracer::startSpan(const Tracing::Config& config, const std::string& operation_name,
+                                   SystemTime start_time,
+                                   const Tracing::Decision tracing_decision) {
+  // Create an Tracers::OpenTelemetry::Span class that will contain the OTel span.
+  Span new_span = Span(config, operation_name, start_time, time_source_, *this);
+  new_span.setSampled(tracing_decision.traced);
+  uint64_t trace_id_high = random_.random();
+  uint64_t trace_id = random_.random();
+  new_span.setTraceId(absl::StrCat(Hex::uint64ToHex(trace_id_high), Hex::uint64ToHex(trace_id)));
+  uint64_t span_id = random_.random();
+  new_span.setId(Hex::uint64ToHex(span_id));
+  return std::make_unique<Span>(new_span);
+}
+
+Tracing::SpanPtr Tracer::startSpan(const Tracing::Config& config, const std::string& operation_name,
+                                   SystemTime start_time,
+                                   const SpanContext& previous_span_context) {
+  // Create a new span and populate details from the span context.
+  Span new_span = Span(config, operation_name, start_time, time_source_, *this);
+  new_span.setSampled(previous_span_context.sampled());
+  new_span.setTraceId(previous_span_context.traceId());
+  if (!previous_span_context.parentId().empty()) {
+    new_span.setParentId(previous_span_context.parentId());
+  }
+  // Generate a new identifier for the span id.
+  uint64_t span_id = random_.random();
+  new_span.setId(Hex::uint64ToHex(span_id));
+  // Respect the previous span's sampled flag.
+  new_span.setSampled(previous_span_context.sampled());
+  return std::make_unique<Span>(new_span);
+}
+
+} // namespace OpenTelemetry
+} // namespace Tracers
+} // namespace Extensions
+} // namespace Envoy
```

Comment:
missing newline - the easiest is to fix your editor to prevent this

Reply:
Ah whoops yes of course, thanks for catching that! Must have missed this in my editor settings.